core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -420,63 +419,81 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, AggregatedResu
+    }
+
+    private HttpResponse convertResponseInternal(ServiceRequestContext ctx,
+                                                 ResponseHeaders headers,
+                                                 @Nullable Object result,
+                                                 HttpHeaders trailers) {
         if (result instanceof CompletionStage) {
             final CompletionStage<?> future = (CompletionStage<?>) result;
-            return HttpResponse.from(future.thenApply(object -> convertResponse(ctx, newHeaders, object,
-                                                                                newTrailers)));
+            return HttpResponse.from(
+                    future.thenApply(object -> convertResponseInternal(ctx, headers, object, trailers)));
         }
 
         try (SafeCloseable ignored = ctx.push()) {
-            return responseConverter.convertResponse(ctx, newHeaders, result, newTrailers);
+            return responseConverter.convertResponse(ctx, headers, result, trailers);
         } catch (Exception cause) {
             return HttpResponse.ofFailure(cause);
         }
     }
 
-    private static ResponseHeadersBuilder addNegotiatedResponseMediaType(ServiceRequestContext ctx,
-                                                                         HttpHeaders headers) {
+    private ResponseHeaders buildResponseHeaders(ServiceRequestContext ctx, HttpHeaders customHeaders) {
+        final ResponseHeadersBuilder builder;
 
-        final MediaType negotiatedResponseMediaType = ctx.negotiatedResponseMediaType();
-        if (negotiatedResponseMediaType == null || headers.contentType() != null) {
-            // Do not overwrite 'content-type'.
-            return ResponseHeaders.builder()
-                                  .add(headers);
+        // Prefer ResponseHeaders#toBuilder because builder#add is an expensive operation.
+        if (customHeaders instanceof ResponseHeaders) {
+            builder = ((ResponseHeaders) customHeaders).toBuilder();
+        } else {
+            builder = ResponseHeaders.builder();
+            builder.add(customHeaders);
+            if (!builder.contains(HttpHeaderNames.STATUS)) {
+                builder.status(defaultStatus);
+            }
         }
+        return addContentTypeIfNeeded(ctx, builder).build();
+    }
 
-        return ResponseHeaders.builder()
-                              .add(headers)
-                              .contentType(negotiatedResponseMediaType);
+    private ResponseHeaders buildResponseHeaders(ServiceRequestContext ctx) {
+        return addContentTypeIfNeeded(ctx, ResponseHeaders.builder(defaultStatus)).build();
     }
 
-    private ResponseHeaders setHttpStatus(ResponseHeadersBuilder headers) {
-        if (headers.contains(HttpHeaderNames.STATUS)) {
-            // Do not overwrite HTTP status.
-            return headers.build();
+    private static ResponseHeadersBuilder addContentTypeIfNeeded(ServiceRequestContext ctx,
```

Comment:
Ignorable) We usually use `maybe` prefix if an operation is performed conditionally. 
```suggestion
    private static ResponseHeadersBuilder maybeAddContentType(ServiceRequestContext ctx,
```